mne/decoding/gsoc.py

```diff
@@ -0,0 +1,523 @@
+        type, else empty.
+
+
+    See Also
+    --------
+    mne.preprocessing.Xdawn
+    CSP
+
+    References
+    ----------
+    [1] Rivet, B., Souloumiac, A., Attina, V., & Gibert, G. (2009). xDAWN
+    algorithm to enhance evoked potentials: application to brain-computer
+    interface. Biomedical Engineering, IEEE Transactions on, 56(8), 2035-2043.
+
+    [2] Rivet, B., Cecotti, H., Souloumiac, A., Maby, E., & Mattout, J. (2011,
+    August). Theoretical analysis of xDAWN algorithm: application to an
+    efficient sensor selection in a P300 BCI. In Signal Processing Conference,
+    2011 19th European (pp. 1382-1386). IEEE.
+    """
+
+    def __init__(self, n_components=2, n_chan=None, signal_cov=None,
+                 reg=None):
+        """init xdawn."""
+        if n_chan is None:
+            raise ValueError("n_chan cannot be none. Please provide the "
+                             "number of channels")
+        self.n_chan = n_chan
+        self.n_components = n_components
+        self.signal_cov = signal_cov
+        self.reg = reg
+
+    def fit(self, X, y):
+        """Fit Xdawn from epochs.
+
+        Parameters
+        ----------
+        X : ndarray, shape(n_channels, n_times * n_freq)
+            Data of epochs.
+        y : ndarray shape(n_samples,)
+            Target values.
+
+        Returns
+        -------
+        self : XdawnTransformer instance
+            The XdawnTransformer instance.
+        """
+        from sklearn.preprocessing import LabelEncoder
+
+        if X.ndim != 2 or not isinstance(X, np.ndarray):
+            raise ValueError("X should be 2 dimensional ndarray")
+        if not isinstance(y, np.ndarray):
+            raise ValueError("Labels must be numpy array")
+
+        epochs_data = X.reshape(X.shape[0], self.n_chan, X.shape[1] /
+                                self.n_chan)
+        # Extract signal covariance
+        if self.signal_cov is None:
+            sig_data = np.hstack(epochs_data)
+            self.signal_cov_ = _regularized_covariance(sig_data, self.reg)
+        elif isinstance(self.signal_cov, Covariance):
+            self.signal_cov_ = self.signal_cov.data
+        elif isinstance(self.signal_cov, np.ndarray):
+            self.signal_cov_ = self.signal_cov
+        else:
+            raise ValueError('signal_cov must be None, a covariance instance '
+                             'or a ndarray')
+
+        # estimates evoked covariance
+        self.filters_ = dict()
+        self.patterns_ = dict()
+        event_id_cov_ = dict()
+        event_id_mean = dict()
+        toeplitz = dict()
+        classes = LabelEncoder().fit(y).classes_
+        shape = epochs_data.shape
+        for eid in classes:
+            mean_data = np.mean(epochs_data[eid].reshape(-1, shape[1],
+                                shape[2]), axis=0)
+            event_id_mean[eid] = mean_data
+            toeplitz[eid] = 1.0
+
+        for eid in classes:
```

Comment:
if classes are list of int, why do we contruct a dict? a numpy array should suffice no?